torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -435,28 +450,77 @@ TensorView* TensorView::cache_before() {
   // Expr* producer_uses =
   new UnaryOp(UnaryOpType::Set, consumer, producer);
 
+  if (consumer_replay_needed) {
+    TransformReplay::replayCasP(consumer, producer, -1);
+  }
+
+  // Make the cache tensor computed at the consumer if the
+  // consumer is computed at another tensor. The position is
+  // the same as this position of the consumer. Note that since
+  // the consumer is computed at another tensor at this position,
+  // there must not be reduction domains in domains until this
+  // position, so the removal of reduction domains should not affect
+  // position indices.
+  // First, make the cache tensor needs look like the consumer. The
+  // minimum number of axes to share is getThisComputeAtAxis(), but
+  // it's safe to fully replay.
+
   // Before: This TV -> Next TV
   // After:  New TV (CB) -> This TV -> Next TV
   if (hasComputeAt()) {
-    TransformReplay::replayPasC(producer, consumer, -1);
-    auto this_ca_pos = getThisComputeAtAxis();
-    producer->computeAt(consumer, this_ca_pos);
-  } else {
-    // Before: Prev TV -> This TV
-    // After:  Prev TV -> New TV (CB) -> This TV
-    // Iterate over origin expression inputs for cache_before on outputs
-    for (TensorView* origin_input :
-         ir_utils::filterByType<TensorView>(expr_inputs)) {
-      if (origin_input->hasComputeAt() &&
-          origin_input->getComputeAtView() == this) {
+    if (!cache_replayed) {
+      TransformReplay::replayPasC(producer, consumer, -1);
+      cache_replayed = true;
+    }
+    producer->setComputeAt(consumer, (int)getThisComputeAtAxis());
+  }
+
+  // If the consumer was the target of computeAt by its input input
+  // tensors, change the computeAt target to the cache
+  // tensor.
+
+  // Before: Prev TV -> This TV
+  // After:  Prev TV -> New TV (CB) -> This TV
+  // Iterate over origin expression inputs for cache_before on outputs
+  auto producer_this_pos = producer->getThisComputeAtAxis();
+  for (TensorView* origin_input :
+       ir_utils::filterByType<TensorView>(expr_inputs)) {
+    if (origin_input->hasComputeAt() &&
+        origin_input->getComputeAtView() == this) {
+      if (!cache_replayed) {
         TransformReplay::replayPasC(producer, consumer, -1);
+        cache_replayed = true;
+      }
+      auto origin_rel_ca_pos = origin_input->getRelativeComputeAtAxis();
+      origin_input->setComputeAt(producer, origin_rel_ca_pos);
```

Comment:
You can use this function instead as this and relative position don't need to be recomputed:
```.cpp
  // Set all computeAt members without checking any correctness. Useful for
  // computeAt with outputs relative to eachother
  void setComputeAt(TensorView* computeAtView, int thisPos, int relPos);
```

Reply:
Thanks. Changed to use that version of setComputeAt.